### PR TITLE
External links re-grouped by available protocols

### PR DESCRIPTION
**Pull request by pmbohm**
A solution for https://github.com/aodn/aodn-portal/issues/1771
Sorts external links according to the currently available protocols;
Natalia has created a report which shows the possible protocols are less numerous than first thought. The email is added

### CODE REVIEW

**Comment by pmbohm** [comment]
All available protocols – things highlighted (bolded) are the only protocols we use.
 
We have inserted manually the following into the XML:
Gogoduck - **IMOS:AGGREGATION--gogoduck**
Bodaac - **IMOS:AGGREGATION--bodaac**
Aodacc - **IMOS:AGGREGATION--aodaac**
 
 
**WWW:LINK-1.0-http--metadata-URL** (auto-generated)
ESRI:AIMS--http--configuration
ESRI:AIMS--http-get-feature
ESRI:AIMS--http-get-image
GLG:KML-2.0-http-get-map
OGC:CSW
OGC:KML
OGC:GML
OGC:ODS
OGC:OGS
OGC:OUS
OGC:OPS
OGC:ORS
OGC:SOS
OGC:SPS
OGC:SAS
OGC:WCS
OGC:WCS-1.1.0-http-get-capabilities
OGC:WCTS
OGC:WFS
**OGC:WFS-1.0.0-http-get-capabilities**
OGC:WFS-G
OGC:WMC-1.1.0-http-get-capabilities
OGC:WMS
OGC:WMS-1.1.1-http-get-capabilities
OGC:WMS-1.3.0-http-get-capabilities
**OGC:WMS-1.1.1-http-get-map**
OGC:WMS-1.3.0-http-get-map
OGC:SOS-1.0.0-http-get-observation
OGC:SOS-1.0.0-http-post-observation
OGC:WNS
OGC:WPS
WWW:DOWNLOAD-1.0-ftp--download
**WWW:DOWNLOAD-1.0-http—download**
FILE:GEO
FILE:RASTER
WWW:LINK-1.0-http--ical
**WWW:LINK-1.0-http—link**
WWW:LINK-1.0-http--related
WWW:LINK-1.0-http--rss
WWW:LINK-1.0-http—samples
DB:POSTGIS
DB:ORACLE
WWW:LINK-1.0-http—opendap
RBNB:DATATURBINE
UKST
 
 
So as discussed – when we had the ANZMEST version – we had several other protocols that helped to separate both the uploaded files and linked pages into type. These protocols are no longer available to us (I have flagged that they would be a useful feature to return – and perhaps this is further evidenced by Kevin’s suggestion for Step 2). Without these additional protocols – there is no way of distinguishing between the types of uploaded files (previously you could tell if it was a data file, or a “other” file), and the same with online lines (previously there was a “data for download URL” which distinguishes data URLs from simple web addresses – e.g. web sites describing projects.
 
So using the current protocols available, and these being the ones that I use in all the 123 metadata, all you could do is the following:
 
Point of truth URL of this metadata record
 
Files For Download
WWW:DOWNLOAD-1.0-http—download
 
Online Resources
WWW:LINK-1.0-http—link
IMOS:AGGREGATION--gogoduck
IMOS:AGGREGATION--bodaac
IMOS:AGGREGATION--aodaac
 
 
 
-----Original Message-----
From: Philip Bohm
Sent: Wednesday, 9 September 2015 3:32 PM
To: Natalia Atkins
Subject: MEST download protocols
 
Nat,
 
I'm after a list of possible mest link protocols eg:
            "WWW:LINK-1.0-http--metadata-URL",
            "WWW:DOWNLOAD-1.0-http--downloadother", **- protocol no longer available**
            "WWW:LINK-1.0-http--link",
            "WWW:DOWNLOAD-1.0-http--download",
            "WWW:DOWNLOAD-1.0-http--downloaddata", **- protocol no longer available**
            "WWW:LINK-1.0-http--downloaddata" **- protocol no longer available**
 
relates to:
https://github.com/aodn/aodn-portal/issues/1771

File: grails-app/conf/Config.groovy [hunk 1]
```diff
@@ -257,18 +257,13 @@ portal {
 
         dataFile = [
             "WWW:DOWNLOAD-1.0-http--download",
-            "WWW:DOWNLOAD-1.0-http--downloaddata",
-            "WWW:DOWNLOAD-1.0-http--downloadother",
-            "WWW:LINK-1.0-http--downloaddata"
+            "IMOS:AGGREGATION--gogoduck",
+            "IMOS:AGGREGATION--bodaac",
+            "IMOS:AGGREGATION--aodaac"
```
Comment (danfruehauf) [inline grails-app/conf/Config.groovy]: Unused too, isn't it?

Comment (pmbohm) [inline grails-app/conf/Config.groovy]: @danfruehauf 
They are referred to in the portal code to get the download links.
BaseInjector.js: 29
```
return dataCollection.getLinksByProtocol(Portal.app.appConfig.portal.metadataProtocols.dataFile);
```
I could be missing something here so let me know :)

Comment (danfruehauf) [inline grails-app/conf/Config.groovy]: AODAAC?

Comment (pmbohm) [inline grails-app/conf/Config.groovy]: Yep thats what Nat has reported to me. Should i tell her anything to fix? No more AODACC?

Comment (danfruehauf) [inline grails-app/conf/Config.groovy]: Hopefully :)

AODAAC hasn't been around for a while.

**Comment by pmbohm** [comment]
@nat now tells me the AODAAC protocol isnt used @danfruehauf

**Comment by danfruehauf** [comment]
@pmbohm You probably tagged the wrong person, it's @atkinsn that you're after.

**Comment by jkburges** [comment]
What's a good collection to test this with?

**Comment by jkburges** [comment]
> What's a good collection to test this with?

Never mind, I tried a couple.  Looks much better.

File: src/test/javascript/portal/details/InfoPanelSpec.js [hunk 2]
```diff
@@ -9,7 +9,7 @@ describe("Portal.details.InfoPanel", function() {
 
     var panel;
     var abstractTitle = OpenLayers.i18n('abstractTitle');
-    var webpageLinksTitle = OpenLayers.i18n('webpageLinksTitle');
+    var supplementaryLinksTitle = OpenLayers.i18n('supplementaryLinksTitle');
```
Comment (jkburges) [inline src/test/javascript/portal/details/InfoPanelSpec.js]: This is a better name, I couldn't think of one when I came up with the very creative `webpage`.

File: web-app/js/portal/data/MetadataRecord.js [hunk 3]
```diff
@@ -126,7 +126,10 @@ Portal.data.MetadataRecord = function() {
 
         var allowedOnlineResources = [];
 
-        Ext.each(Portal.app.appConfig.portal.metadataProtocols.webPage, function(protocol) {
+        Ext.each(Portal.app.appConfig.portal.metadataProtocols.supplementary, function(protocol) {
+            allowedOnlineResources.push(protocol.trim());
+        });
+        Ext.each(Portal.app.appConfig.portal.metadataProtocols.metadataRecord, function(protocol) {
```
Comment (jkburges) [inline web-app/js/portal/data/MetadataRecord.js]: `metadataRecord` wasn't included before, why now?

Also, the two `Ext.each()` can probably be rolled in to one:

```javascript
Ext.each(Portal.app.appConfig.portal.metadataProtocols.supplementary.concat(Portal.app.appConfig.portal.metadataProtocols.metadataRecord), function(protocol) {
  allowedOnlineResources.push(protocol.trim());
});

File: web-app/js/portal/details/InfoPanel.js [hunk 4]
```diff
@@ -27,27 +28,27 @@ Portal.details.InfoPanel = Ext.extend(Ext.Container, {
...
-            this._getHtmlForLinks()
-        );
+            this._getMetadataLink(),
+            OpenLayers.i18n('supplementaryLinksTitle'),
+            this._getHtmlsupplementaryLinks()
+            );
     },
 
-    _getHtmlForLinks: function() {
+    _getHtmlsupplementaryLinks: function() {
```
Comment (jkburges) [inline web-app/js/portal/details/InfoPanel.js]: `_getHtmlForSupplementaryLinks()`?

Comment (jkburges) [inline web-app/js/portal/details/InfoPanel.js]: Actually, I think `_getSupplementaryLinkAsHtml()` would be a better name.
And, I think also rename `_getMetadataLink()` to `_getMetadataLinkAsHtml()`, for consistency.

**Comment by jkburges** [comment]
@pmbohm just a reminder, there are a couple of comments on here.

File: web-app/js/portal/data/MetadataRecord.js (hunk 3 above)
Comment (pmbohm) [inline web-app/js/portal/data/MetadataRecord.js]: `_getPointOfTruthLink` in baseInjector.js uses `metadataRecord` 
`_getMetadataLink` calls `_getPointOfTruthLink`

**Comment by pmbohm** [comment]
@jkburges New pull request up 
thanks

**Comment by pmbohm** [comment]
@jkburges New pull request up - working

**Comment by jkburges** [comment]
@pmbohm, getting the following error when selecting a collection:

```
Uncaught TypeError: Cannot read property 'indexOf' of undefinedmatchesProtocols @ DataCollection.js:99
```

:confused:

**Comment by pmbohm** [comment]
It might require Config.groovy reloading. IE: restart

**Comment by jkburges** [comment]
> It might require Config.groovy reloading. IE: restart

Apologies, not sure what was going on at my end, but working now.

**Comment by atkinsn** [comment]
I am now going through and adding links to the FTP site for advanced users to access large amounts of data. I have done this in production on the Argo records, however this link does not show up in Step 2 of the Portal under the info tab. I believe this is due to the fact that you have a defined list of recognised protocols - and this is not one of them: WWW:DOWNLOAD-1.0-ftp--download. My concern going forward is with records supplied by our AODN partners, is that we can't possibly know what protocols they might be using (unlikely to have too many outside the "standard" ones however), and we run the risk of excluding links in Step 2 without the contributor realising this. Would it not be better to still display unrecognised links at the bottom - which will then be flagged as standing out when Kate does some testing - and we can then work to add the protocol to a category?

**Pull request merged**